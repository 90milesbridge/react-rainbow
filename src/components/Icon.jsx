import React from 'react';
import classnames from 'classnames';
import Config from './config';

function getIconSprite(name) {
    return name.split(':')[0];
}

function getIconName(name) {
    return name.split(':')[1];
}

export default class Icon extends React.Component {
    render() {
        return (
            <span className={ this.getContainerClasses() } title={ this.props.assistiveText }>
                <IconSvg iconName={ this.props.iconName } iconSize={ this.props.iconSize }/>
                <span className="slds-assistive-text">{ this.props.assistiveText }</span>
            </span>
        );
    }

    getContainerClasses() {
        return classnames('slds-icon_container', `slds-icon-${ getIconSprite(this.props.iconName) }-${ getIconName(this.props.iconName) }`);
    }
}

Icon.propTypes = {
    iconName: React.PropTypes.string.isRequired
};

export class IconSvg extends React.Component {
    render() {
        let pathToAssets = Config.get('pathToAssets');
        let useTag = `<use xlink:href="${ pathToAssets }/icons/${ getIconSprite(this.props.iconName) }-sprite/svg/symbols.svg#${ getIconName(this.props.iconName ) }"></use>`;
        return <svg aria-hidden="true" className={ this.getSvgClasses.call(this) } dangerouslySetInnerHTML={{__html: useTag }}/>
    }

    getSvgClasses() {
        return classnames(this.props.className, {
            'slds-icon': !this.props.className,
            'slds-icon-text-default': getIconSprite(this.props.iconName) === 'utility' && !this.props.className,
            'slds-icon--x-small': this.props.iconSize === 'x-small',
            'slds-icon--small': this.props.iconSize === 'small',
            'slds-icon--large': this.props.iconSize === 'large'
        });
    }
}

