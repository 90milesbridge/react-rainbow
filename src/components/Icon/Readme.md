action icons

    <div className="slds-p-vertical_large slds-p-left_medium slds-grid slds-media_center slds-wrap">
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="action:add_photo_video"
                assistiveText="camera icon"
                title="camera icon"
                size="large" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="action:delete"
                assistiveText="delete icon"
                title="delete icon"
                size="medium" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon 
                iconName="action:description"
                assistiveText="description icon" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon 
                iconName="action:announcement"
                assistiveText="announcement icon"
                size="small" />

        </div>

    </div>

custom icons

    <div className="slds-p-vertical_large slds-p-left_medium slds-grid slds-media_center slds-wrap">
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="custom:custom1"
                assistiveText="heart icon"
                title="heart icon"
                size="large" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="custom:custom5"
                assistiveText="tree icon"
                size="medium" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="custom:custom22"
                assistiveText="phone icon"
                title="phone icon" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="custom:custom108"
                assistiveText="settings icon"
                size="small" />

        </div>
    </div>

doctype icons

    <div className="slds-p-vertical_large slds-p-left_medium slds-grid slds-media_center slds-wrap">
        <div className="slds-m-horizontal_medium"> 
            <Icon
                iconName="doctype:ai"
                assistiveText="ai icon"
                title="ai icon"
                size="large" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="doctype:attachment"
                assistiveText="attachment icon"
                title="attachment icon"
                size="medium" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon iconName="doctype:html"
                assistiveText="html icon"
                title="html icon" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="doctype:quip_sheet"
                assistiveText="quip sheet icon"
                title="quip sheet icon"
                size="small" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon 
                iconName="doctype:link"
                assistiveText="link icon"
                title="link icon"
                size="x-small" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon 
                iconName="doctype:folder"
                assistiveText="folder icon"
                title="folder icon"
                size="xx-small" />

        </div>
    </div>

standard icons

    <div className="slds-p-vertical_large slds-p-left_medium slds-grid slds-media_center slds-wrap">
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="standard:account"
                assistiveText="account icon"
                title="account icon"
                size="large" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="standard:address"
                assistiveText="address icon"
                title="address icon"
                size="medium" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="standard:asset_relationship"
                assistiveText="relationship icon" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon 
                iconName="standard:coaching"
                assistiveText="coaching icon"
                size="small" />

        </div>
    </div>

default utility icons

    <div className="slds-p-vertical_large slds-p-left_medium slds-grid slds-media_center slds-wrap">
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="utility:activity"
                assistiveText="activity icon"
                title="activity icon"
                size="large" />
        
        </div>
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="utility:ad_set"
                assistiveText="ad set icon"
                size="medium" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon 
                iconName="utility:bucket"
                assistiveText="bucket icon" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="utility:answer"
                assistiveText="answer icon"
                title="answer icon"
                size="small" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="utility:anchor"
                assistiveText="anchor icon"
                title="anchor icon"
                size="x-small" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="utility:like"
                assistiveText="address icon"
                size="xx-small" />

        </div>
    </div>

utility icons variants

    <div className="slds-p-vertical_large slds-p-left_medium slds-grid slds-media_center slds-wrap">
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="utility:warning"
                title="warning icon"
                variant="warning"
                size="large" />
        
        </div>
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="utility:ban"
                assistiveText="ban icon"
                title="ban icon"
                variant="error"
                size="medium" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="utility:help"
                assistiveText="help icon"
                variant="light" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="utility:dash"
                assistiveText="dash icon"
                title="dash icon"
                size="small" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="utility:add"
                assistiveText="add icon"
                size="x-small"
                variant="light" />

        </div>
        <div className="slds-m-horizontal_medium">
            <Icon
                iconName="utility:power"
                assistiveText="power icon"
                size="xx-small" />

        </div>
    </div>