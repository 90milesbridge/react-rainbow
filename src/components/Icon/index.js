import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import iconNamePropType from './../../propTypes/iconNamePropType';
import IconSvg from './../IconSvg';
import AssistiveText from './../AssistiveText';
import {
    isActionSprite,
    isUtilitySprite,
    normalizeIconName,
    hasBackgroundColor,
} from './utils';

/**
* The Icon component is the Lightning Design System Icon
* component and should be used for naked icons.
*/
export default function Icon(props) {
    const {
        className,
        svgClassName,
        style,
        svgStyle,
        iconName,
        title,
        assistiveText,
        size,
        variant,
    } = props;

    const getContainerClassNames = () => {
        const icon = normalizeIconName(iconName);
        const iconBackgroundColorClassName = `slds-icon-${icon}`;

        return classnames('slds-icon_container', {
            'slds-icon_container_circle': isActionSprite(iconName),
            [iconBackgroundColorClassName]: hasBackgroundColor(iconName),
        }, className);
    };

    const getSvgSizeClassName = () => {
        const isMediumSize = size === 'medium';
        const isXSmall = size === 'x-small' || size === 'xx-small';
        if (isMediumSize) {
            return '';
        }
        if (hasBackgroundColor(iconName) && isXSmall) {
            return 'slds-icon_small';
        }
        return `slds-icon_${size}`;
    };

    const getSvgClassNames = () => classnames('slds-icon', getSvgSizeClassName(), {
        [`slds-icon-text-${variant}`]: isUtilitySprite(iconName),
    }, svgClassName);

    return (
        <span className={getContainerClassNames()} title={title} style={style}>
            <IconSvg className={getSvgClassNames()} style={svgStyle} iconName={iconName} />
            <AssistiveText text={assistiveText} />
        </span>
    );
}

Icon.propTypes = {
    /** The Lightning Design System name of the icon used as a fallback when
    * the image fails to load. Names are written in the format {sprite_name}:{icon_name}
    * where {sprite_name} is the category, and {icon_name} is the specific icon to be displayed.
    * The icon is displayed in the header to the left of the title. */
    iconName: iconNamePropType.simple.isRequired,
    /** The size of the icon. Options include xx-small, x-small, small, medium, or large.
    * This value defaults to medium. */
    size: PropTypes.oneOf([
        'xx-small',
        'x-small',
        'small',
        'medium',
        'large',
    ]),
    /** The variant changes the appearance of a utility icon.
    * Accepted variants include default, light, warning, and error.
    * Use the light variant to implement a white fill in utility icons on dark backgrounds. */
    variant: PropTypes.oneOf([
        'default',
        'light',
        'error',
        'warning',
    ]),
    /** Displays tooltip text when the mouse moves over the element. */
    title: PropTypes.string,
    /** A description for assistive sreen readers. */
    assistiveText: PropTypes.string,
    /** It is an object with custom style applied to the svg element. */
    svgStyle: PropTypes.object,
    /** A CSS class for the outer element, in addition to the component's base classes. */
    className: PropTypes.string,
    /** A CSS class for the  svg element. */
    svgClassName: PropTypes.string,
    /** An object with custom style applied to the outer element. */
    style: PropTypes.object,
};

Icon.defaultProps = {
    className: undefined,
    svgClassName: undefined,
    style: undefined,
    svgStyle: undefined,
    title: undefined,
    assistiveText: undefined,
    size: 'medium',
    variant: 'default',
};
