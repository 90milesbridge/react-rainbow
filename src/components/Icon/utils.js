const UTILITY_REGEX = /^utility:.+/;
const ACTION_REGEX = /^action:.+/;
const CUSTOM_REGEX = /^custom:.+/;
const STANDARD_REGEX = /^standard:.+/;

export function isUtilitySprite(iconName) {
    return UTILITY_REGEX.test(iconName);
}

export function isActionSprite(iconName) {
    return ACTION_REGEX.test(iconName);
}

export function isCustomSprite(iconName) {
    return CUSTOM_REGEX.test(iconName);
}

export function isStandardSprite(iconName) {
    return STANDARD_REGEX.test(iconName);
}

export function normalizeIconName(iconName) {
    return iconName.replace(/_/g, '-').replace(':', '-');
}

export function hasBackgroundColor(iconName) {
    return isActionSprite(iconName)
        || isCustomSprite(iconName)
        || isStandardSprite(iconName);
}
