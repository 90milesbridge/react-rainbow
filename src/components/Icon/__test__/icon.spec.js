import React from 'react';
import { shallow } from 'enzyme';
import Icon from './../index';

describe('<Icon/>', () => {
    it(': should be replaced by -', () => {
        const component = shallow(<Icon iconName="custom:custom5" />);

        expect(component.find('.slds-icon-custom-custom5').exists()).toBe(true);
    });
});
