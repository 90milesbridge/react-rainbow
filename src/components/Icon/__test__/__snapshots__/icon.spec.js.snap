// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon/> should have the right class names when custom className and svgClassName are passed 1`] = `
<span
  className="slds-icon_container my-custom-class-name"
>
  <svg
    aria-hidden={true}
    className="slds-icon slds-icon-text-default my-custom-svg-class-name"
  >
    <use
      xlinkHref="/icons/utility-sprite/svg/symbols.svg#adduser"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</span>
`;

exports[`<Icon/> should have the right class names when size and variant are not passed 1`] = `
<span
  className="slds-icon_container slds-icon-custom-custom5"
>
  <svg
    aria-hidden={true}
    className="slds-icon"
  >
    <use
      xlinkHref="/icons/custom-sprite/svg/symbols.svg#custom5"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</span>
`;

exports[`<Icon/> should have the right class names when size is large 1`] = `
<span
  className="slds-icon_container slds-icon_container_circle slds-icon-action-add-photo-video"
>
  <svg
    aria-hidden={true}
    className="slds-icon slds-icon_large"
  >
    <use
      xlinkHref="/icons/action-sprite/svg/symbols.svg#add_photo_video"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</span>
`;

exports[`<Icon/> should have the right class names when size is medium 1`] = `
<span
  className="slds-icon_container slds-icon-standard-action-list-component"
>
  <svg
    aria-hidden={true}
    className="slds-icon"
  >
    <use
      xlinkHref="/icons/standard-sprite/svg/symbols.svg#action_list_component"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</span>
`;

exports[`<Icon/> should have the right class names when size is small 1`] = `
<span
  className="slds-icon_container slds-icon-custom-custom6"
>
  <svg
    aria-hidden={true}
    className="slds-icon slds-icon_small"
  >
    <use
      xlinkHref="/icons/custom-sprite/svg/symbols.svg#custom6"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</span>
`;

exports[`<Icon/> should have the right class names when size is x-small 1`] = `
<span
  className="slds-icon_container"
>
  <svg
    aria-hidden={true}
    className="slds-icon slds-icon_x-small"
  >
    <use
      xlinkHref="/icons/doctype-sprite/svg/symbols.svg#box_notes"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</span>
`;

exports[`<Icon/> should have the right class names when size is x-small and the sprite is custom 1`] = `
<span
  className="slds-icon_container slds-icon-custom-custom5"
>
  <svg
    aria-hidden={true}
    className="slds-icon slds-icon_small"
  >
    <use
      xlinkHref="/icons/custom-sprite/svg/symbols.svg#custom5"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</span>
`;

exports[`<Icon/> should have the right class names when size is xx-small 1`] = `
<span
  className="slds-icon_container"
>
  <svg
    aria-hidden={true}
    className="slds-icon slds-icon_xx-small slds-icon-text-default"
  >
    <use
      xlinkHref="/icons/utility-sprite/svg/symbols.svg#ad_set"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</span>
`;

exports[`<Icon/> should have the right class names when size is xx-small and the sprite is action 1`] = `
<span
  className="slds-icon_container slds-icon_container_circle slds-icon-action-add-contact"
>
  <svg
    aria-hidden={true}
    className="slds-icon slds-icon_small"
  >
    <use
      xlinkHref="/icons/action-sprite/svg/symbols.svg#add_contact"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</span>
`;

exports[`<Icon/> should have the right class names when variant is default 1`] = `
<span
  className="slds-icon_container"
>
  <svg
    aria-hidden={true}
    className="slds-icon slds-icon-text-default"
  >
    <use
      xlinkHref="/icons/utility-sprite/svg/symbols.svg#animal_and_nature"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</span>
`;

exports[`<Icon/> should have the right class names when variant is error 1`] = `
<span
  className="slds-icon_container"
>
  <svg
    aria-hidden={true}
    className="slds-icon slds-icon-text-error"
  >
    <use
      xlinkHref="/icons/utility-sprite/svg/symbols.svg#approval"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</span>
`;

exports[`<Icon/> should have the right class names when variant is light 1`] = `
<span
  className="slds-icon_container"
>
  <svg
    aria-hidden={true}
    className="slds-icon slds-icon-text-light"
  >
    <use
      xlinkHref="/icons/utility-sprite/svg/symbols.svg#answer"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</span>
`;

exports[`<Icon/> should have the right class names when variant is passed and the sprite is other than utility 1`] = `
<span
  className="slds-icon_container slds-icon_container_circle slds-icon-action-add-contact"
>
  <svg
    aria-hidden={true}
    className="slds-icon"
  >
    <use
      xlinkHref="/icons/action-sprite/svg/symbols.svg#add_contact"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</span>
`;

exports[`<Icon/> should have the right class names when variant is warning 1`] = `
<span
  className="slds-icon_container"
>
  <svg
    aria-hidden={true}
    className="slds-icon slds-icon-text-warning"
  >
    <use
      xlinkHref="/icons/utility-sprite/svg/symbols.svg#approval"
      xmlnsXlink="http://www.w3.org/1999/xlink"
    />
  </svg>
</span>
`;
